contracts/stake/StakedTokenV3.sol

```diff
@@ -364,6 +366,41 @@ contract StakedTokenV3 is StakedTokenV2, IStakedTokenV3, RoleManager {
     emit MaxSlashablePercentageChanged(percentage);
   }
 
+  /**
+   * @dev Snapshots the current exchange rate
+   */
+  function snapshotExchangeRate() public {
+    uint128 currentBlock = uint128(block.number);
+    uint128 newExchangeRate = uint128(exchangeRate());
+
+    // Doing multiple operations in the same block
+    if (
+      _countExchangeRateSnapshots != 0 &&
+      _exchangeRateSnapshots[_countExchangeRateSnapshots - 1].blockNumber == currentBlock
+    ) {
+      _exchangeRateSnapshots[_countExchangeRateSnapshots - 1].value = newExchangeRate;
+    } else {
+      _exchangeRateSnapshots[_countExchangeRateSnapshots] = Snapshot(currentBlock, newExchangeRate);
+      _countExchangeRateSnapshots = _countExchangeRateSnapshots + 1;
+    }
+    emit EchangeRateSnapshotted(currentBlock, newExchangeRate);
+  }
+
+  /**
+   * @dev Calculates the exchange rate between the amount of STAKED_TOKEN and the the StakeToken total supply.
+   * Slashing will reduce the exchange rate. Supplying STAKED_TOKEN to the stake contract
+   * can replenish the slashed STAKED_TOKEN and bring the exchange rate back to 1
+   **/
+  function exchangeRate() public view override returns (uint256) {
+    uint256 currentSupply = totalSupply();
+
+    if (currentSupply == 0) {
+      return 1e18; //initial exchange rate is 1:1
```

Comment:
I think we should use constants for the precision, not literal numbers. For example `EXCHANGE_RATE_PRECISION` or something in that line. Applying it also on the `mul` and `div` calculations using the exchange rate @dhadrien